wp-content/plugins/core/src/Components/Component.php

```diff
@@ -0,0 +1,184 @@
+<?php
+
+namespace Tribe\Project\Components;
+
+use Tribe\Libs\Utils\Markup_Utils;
+use DI;
+use Tribe\Project\Assets\Theme\Theme_Build_Parser;
+use Tribe\Project\Components\Component_Factory;
+use Tribe\Project\Components\Handler;
+use Tribe\Project\Models\Model;
+use Tribe\Project\Models\Post;
+use Twig\Environment;
+use Twig\Error\Error;
+
+abstract class Component {
+
+	/**
+	 * @var bool
+	 */
+	protected $debug = false;
+
+	/**
+	 * @var bool
+	 */
+	protected $truncate_debug_models = true;
+
+	/**
+	 * @var array
+	 */
+	protected $data;
+
+	/**
+	 * @var Component_Factory
+	 */
+	protected $factory;
+
+	/**
+	 * @var Environment
+	 */
+	protected $twig;
+
+	/**
+	 * @var Theme_Build_Parser
+	 */
+	protected $build_parser;
+
+	/**
+	 * @var string
+	 */
+	protected $twig_file = '';
+
+	public function __construct( Component_Factory $factory, Environment $twig, Theme_Build_Parser $build_parser, $args = [] ) {
+		$this->factory      = $factory;
+		$this->twig         = $twig;
+		$this->build_parser = $build_parser;
+
+		$this->data = $args;
+		$this->merge_defaults();
+	}
+
+	private function merge_defaults() {
+		$this->data = wp_parse_args( $this->data, $this->defaults() );
+	}
+
+	public function data(): array {
+		return $this->data;
+	}
+
+	public function render(): void {
+		$twig_file = $this->get_twig_file_for_component();
+
+		include( $twig_file );
+	}
+
+	protected function get_twig_file_for_component() {
+		$dir = dirname( ( new \ReflectionClass( static::class ) )->getFileName() );
+
+		if ( ! empty( $this->twig_file ) ) {
+			return sprintf( '%s/%s', $dir, $this->twig_file );
+		}
+
+		$files = scandir( $dir );
+		$twigs = array_filter( $files, function ( $file ) {
+			$ext = pathinfo( $file, PATHINFO_EXTENSION );
+
+			return $ext === 'twig';
+		} );
+
+		if ( empty( $twigs ) ) {
+			throw new \ErrorException( 'Could not locate a valid Twig template for this component. Please ensure a .twig file exists in the same directory, or that you explicitely define the $twig_file property for this component.' );
+		}
+
+		return sprintf( '%s/%s', $dir, basename( array_shift( $twigs ) ) );
```

Comment:
I noticed while in development, that this method will pull the first twig file which starts with the same string as the component file.

For example, if I have a `Link.php` component with 2 twig files in the same directory: http://p.tri.be/upSuSF
* `Link.php`
* `link.twig`
* `link-with-wrapper.twig`

By default this method is returning `link-with-wrapper.twig` instead of `link.twig`. I think the default template file probably needs to be a direct file name match to the PHP file rather than the first match in the array of twig files.